### DISCUSSION TITLE
Velocity is not evolving in advection equation

### DISCUSSION THREAD

**Discussion by Swarnali-D** [CliMA/Oceananigans.jl#3941] [Computational science!]
I have set up a simple run to test the kinetic energy evolution through Burger's equation. 
However, the numerical solution is not matching the analytical solution. Please let me know what I'm missing 

```
using Oceananigans, CairoMakie

Lx = 1.0  
Nx = 100   
grid = RectilinearGrid(CPU(), 
                     size=(Nx, 4, 4),  
                     x=(0, Lx),
                     y=(0, 0.1),
                     z=(0, 0.1),
                     topology=(Periodic, Periodic, Periodic))


# Initial condition: u(x,0) = 4x
α = 4.0
function u_initial(x, y, z)
    return α*x
end

model = NonhydrostaticModel(
    grid = grid,
    closure = nothing,  
    timestepper = :RungeKutta3,
    advection = Centered(order=2)  
)

set!(model, u=u_initial)

simulation = Simulation(model, Δt=0.001, stop_time=0.15)
u = model.velocities.u
simulation.output_writers[:fields] = JLD2OutputWriter(model, (; u),
                                                    filename = "burgers_adv",
                                                    schedule = TimeInterval(0.01),
                                                    overwrite_existing = true)

run!(simulation)
```
![burgers](https://github.com/user-attachments/assets/69f4ebdf-7f71-4c7c-98e5-2ff388816e55)

**Comment by simone-silvestri** [comment]
The Burgers equation is a "compressible" equation that does not involve a pressure gradient (approximating high Mach number flows). The Navier-Stokes equations, solved in the `NonhydrostaticModel, are incompressible, where a pressure gradient ensures the incompressibility. Therefore, if you initialize the model with a divergent velocity field (as the one you are prescribing), the pressure solver will correct the model to make it incompressible. You could, in principle, adapt the code to solve the burgers equations by removing the pressure solver, but it requires a bit of fiddling with the source.

**Reply by simone-silvestri** [reply]
Also, note that the Burgers equation leads to the formation of shocks, so you should solve them with shock-capturing advection methods such as WENO.

**Comment by glwagner** [comment]
The initial condition $u = \alpha x$ is non-periodic and purely divergent, since $\partial_x u = \alpha$ cannot be balanced by any $\partial_y v$ or $\partial_z w$. As a result, when the divergent component of the velocity field is removed during `set!`, only the horizontal mean of $u$ remains. You can check this without even running the simulation, just plot $u$ after calling `set!`.

**Comment by Swarnali-D** [comment]
If I set `u0(x, y, z) = sin(2π*x)`, my initial condition is still not properly set 
![initial_u](https://github.com/user-attachments/assets/d55aa1ac-7d47-447c-9320-ee588457e593)

**Reply by glwagner** [reply]
The proposed condition is also divergent. What are you looking for?

To say this in a different way --- the `NonhydrostaticModel` can only simulate incompressible velocity fields that are divergence-free. This means that we _enforce_ the condition $\partial_x u + \partial_y v + \partial_z w = 0$.

You could try 

```julia
u0(x, y, z) = sin(2π*y)
```

for example.

Another possibility is to generate a streamfunction $\psi(x, y)$ which is guaranteed to be incompressible, and then generate the velocity field from it by defining $u = \partial_y \psi$ and $v = - \partial_x \psi$.

**Reply by glwagner** [reply]
I'm not sure the best reference but wikipedia might be a good place to start: https://en.wikipedia.org/wiki/Incompressible_flow
